// esbuild-visualizer/dist/plugin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.visualizer = void 0;
const version_1 = require("./version");
const module_mapper_1 = require("./module-mapper");
const data_1 = require("./data");
const build_stats_1 = require("./build-stats");
const visualizer = async (metadata, opts = {}) => {
    var _a, _b;
    const title = (_a = opts.title) !== null && _a !== void 0 ? _a : "EsBuild Visualizer";
    const template = (_b = opts.template) !== null && _b !== void 0 ? _b : "treemap";
    const projectRoot = "";
    const renderedModuleToInfo = (id, mod) => {
        const result = {
            id,
            renderedLength: mod.bytesInOutput,
        };
        return result;
    };
    const roots = [];
    const mapper = new module_mapper_1.ModuleMapper(projectRoot);
    const links = [];
    // collect trees
    for (const [bundleId, bundle] of Object.entries(metadata.outputs)) {
        const modules = Object.entries(bundle.inputs).map(([id, mod]) => renderedModuleToInfo(id, mod));
        const tree = data_1.buildTree(bundleId, modules, mapper);
        roots.push(tree);
    }
    const getModuleInfo = (bundle) => (moduleId) => {
        var _a, _b, _c, _d;
        const input = (_a = metadata.inputs) === null || _a === void 0 ? void 0 : _a[moduleId];
        const imports = input === null || input === void 0 ? void 0 : input.imports.map((i) => i.path);
        return {
            renderedLength: (_d = (_c = (_b = bundle.inputs) === null || _b === void 0 ? void 0 : _b[moduleId]) === null || _c === void 0 ? void 0 : _c.bytesInOutput) !== null && _d !== void 0 ? _d : 0,
            importedIds: imports !== null && imports !== void 0 ? imports : [],
            dynamicallyImportedIds: [],
            isEntry: bundle.entryPoint === moduleId,
            isExternal: false,
        };
    };
    for (const [bundleId, bundle] of Object.entries(metadata.outputs)) {
        if (bundle.entryPoint == null)
            continue;
        data_1.addLinks(bundleId, bundle.entryPoint, getModuleInfo(bundle), links, mapper);
    }
    const tree = data_1.mergeTrees(roots);
    const data = {
        version: version_1.version,
        tree,
        nodes: mapper.getNodes(),
        nodeParts: mapper.getNodeParts(),
        links,
        env: {},
        options: {
            gzip: false,
            brotli: false,
            sourcemap: false,
        },
    };
    const fileContent = await build_stats_1.buildHtml({
        title,
        data,
        template,
    });
    return fileContent;
};
exports.visualizer = visualizer;
